### PR TITLE
Add metricbeat iis module

### PR DESCRIPTION
**Pull request by narph**
Related  issue https://github.com/elastic/beats/issues/13072

Reusing the pdh query (in perfmon) to retrieve the iis specific counters.
Initial implementation contains 2 metricsets:
- webserver - includes aggregated metrics over the entire IIS web server (light metricset based on the perfmon metricset)
- website - includes metrics from each website running 
- application_pool - includes metrics for each application pool

```
- module: iis
  metricsets:
    - webserver
    - website
    - application_pool
  enabled: true
  period: 10s

 # filter on application pool names
 # application_pool.name: []
```

### CODE REVIEW

File: metricbeat/module/windows/perfmon/perfmon.go [hunk 1]
```diff
@@ -63,6 +68,12 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 	if err := base.Module().UnpackConfig(&config); err != nil {
 		return nil, err
 	}
+	if metricsetName == base.Name() {
+		err := validateCounterConfig(config.CounterConfig)
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (jsoriano) [inline metricbeat/module/windows/perfmon/perfmon.go]: Why is this change needed?

File: metricbeat/module/windows/perfmon/reader.go [hunk 2]
```diff
@@ -94,21 +95,20 @@ func NewReader(config Config) (*Reader, error) {
 			r.measurement[v] = counter.MeasurementLabel
 		}
 	}
-
 	return r, nil
 }
 
 // RefreshCounterPaths will recheck for any new instances and add them to the counter list
-func (r *Reader) RefreshCounterPaths() error {
+func (this *Reader) RefreshCounterPaths() error {
```
Comment (jsoriano) [inline metricbeat/module/windows/perfmon/reader.go]: Oh, memories of java world :slightly_smiling_face:

File: metricbeat/module/windows/perfmon/reader.go [hunk 3]
```diff
@@ -219,3 +271,27 @@ func matchesParentProcess(instanceName string) (bool, string) {
 	}
 	return false, instanceName
 }
+
+// filterValuesByInstances func filters the counter valus based on the instances provided by the user, this is applied to the 'website' metricset at the moment so the _Total instance should be filtered out
```
Comment (jsoriano) [inline metricbeat/module/windows/perfmon/reader.go]:
```suggestion
// filterValuesByInstances func filters the counter values based on the instances provided by the user, this is applied to the 'website' metricset at the moment so the _Total instance should be filtered out
```

File: metricbeat/module/windows/perfmon/reader.go [hunk 4]
```diff
@@ -137,33 +137,50 @@ func (r *Reader) RefreshCounterPaths() error {
...
+	values, err := this.query.GetFormattedCounterValues()
 	if err != nil {
 		return nil, errors.Wrap(err, "failed formatting counter values")
 	}
+	var events []mb.Event
+	if metricsetName != metricset {
+		// will be enabled when https://github.com/elastic/beats/issues/16366
+		//if len(this.config.FilterByInstance) > 0 {
+		//	values = filterValuesByInstances(this.config.FilterByInstance, values)
+		//}
```
Comment (jsoriano) [inline metricbeat/module/windows/perfmon/reader.go]: We could still add this filtering as something generic for the perfmon module, wdyt about adding this in a different PR?

File: metricbeat/module/windows/perfmon/reader.go [hunk 5]
```diff
@@ -137,33 +137,50 @@ func (r *Reader) RefreshCounterPaths() error {
...
 	}
 
 	// Get the values.
-	values, err := r.query.GetFormattedCounterValues()
+	values, err := this.query.GetFormattedCounterValues()
 	if err != nil {
 		return nil, errors.Wrap(err, "failed formatting counter values")
 	}
+	var events []mb.Event
+	if metricsetName != metricset {
```
Comment (jsoriano) [inline metricbeat/module/windows/perfmon/reader.go]: Why this difference?

File: metricbeat/helper/windows/pdh/pdh_windows.go [hunk 6]
```diff
@@ -50,6 +51,8 @@ type PdhCounterHandle uintptr
 
 var InvalidCounterHandle = ^PdhCounterHandle(0)
 
+const PERF_DETAIL_WIZARD = 400
```
[the remark this answers is not in the document]
Comment (narph) [reply metricbeat/helper/windows/pdh/pdh_windows.go]: done, thank you

File: metricbeat/docs/modules/iis.asciidoc [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+This is the iis module.
+
+IIS (Internet Information Services) is a secure, reliable, and scalable Web server that provides an easy to manage platform for developing and hosting Web applications and services.
+
+The `iis` module will periodically retrieve IIS related metrics using performance counters such as:
+
+ - System/Process counters like the the overall server and CPU usage for the IIS Worker Process and memory (currently used and available memory for the IIS Worker Process).
+ - IIS performance counters like Web Service: Bytes Received/Sec, Web Service: Bytes Sent/Sec, etc, which are helpful to track to identify potential spikes in traffic.
+ - Web Service Cache counters in order to monitor user mode cache and output cache.
+
```
[the remark this answers is not in the document]
Comment (narph) [reply metricbeat/docs/modules/iis.asciidoc]: added a short description under the listing of the metricsets

File: metricbeat/module/iis/application_pool/application_pool.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The iis application_pool metricset is beta.")
+	var config Config
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+	// instantiate reader object
+	reader, err := newReader()
```
[the remark this answers is not in the document]
Comment (narph) [reply metricbeat/module/iis/application_pool/application_pool.go]: done, thanks

File: metricbeat/module/iis/application_pool/reader.go [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+		for _, w3 := range applicationPools {
+			if n == w3.Name {
+				filtered = append(filtered, w3)
+			}
+		}
+	}
+	return filtered, nil
+}
+
+// getw3wpProceses func retrieves the running w3wp process ids
```
[the remark this answers is not in the document]
Comment (narph) [reply metricbeat/module/iis/application_pool/reader.go]: added more info in the comments of the function.

File: metricbeat/module/iis/application_pool/reader.go [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+		info, err := p.Info()
+		if err != nil {
+			continue
+		}
+		if info.Name == "w3wp.exe" {
+			if len(info.Args) > 0 {
+				for i, ar := range info.Args {
+					if ar == "-ap" && len(info.Args) > i+1 {
+						wps[info.PID] = info.Args[i+1]
+						continue
```
[the remark this answers is not in the document]
Comment (narph) [reply metricbeat/module/iis/application_pool/reader.go]: fixed, thanks for pointing that out

File: metricbeat/module/windows/perfmon/perfmon.go (hunk 1 above)
Comment (narph) [reply metricbeat/module/windows/perfmon/perfmon.go]: the validateCounterConfig func will check if the instanceLabel has been set, the instance label is mandatory inside the pefrmon metricset but not in the webserver metricset

File: metricbeat/module/windows/perfmon/reader.go (hunk 2 above)
Comment (narph) [reply metricbeat/module/windows/perfmon/reader.go]: all changed

File: metricbeat/module/windows/perfmon/reader.go (hunk 4 above)
Comment (narph) [reply metricbeat/module/windows/perfmon/reader.go]: yes, might be something to think about, I have removed the checks for now

File: metricbeat/module/windows/perfmon/reader.go (hunk 5 above)
Comment (narph) [reply metricbeat/module/windows/perfmon/reader.go]: // GroupAllCountersTo config option will only apply to the webserver metricset, where counters for all instances are aggregated, I do not plan to make this option available to the perfmon metricset yet

**Comment by narph** [comment]
@jsoriano , I have touched on all the feedback, let me know if I miss anything

File: metricbeat/module/windows/perfmon/perfmon.go [hunk 11]
```diff
@@ -40,13 +40,18 @@ type CounterConfig struct {
...
+	GroupAllCountersTo string          `config:"perfmon.group_all_counter"` // only available for the iis/webserver metricset at the moment
+	// could be used in the future if any solutions are built for https://github.com/elastic/beats/issues/16366
+	//FilterByInstance   []string        `config:"perfmon.filter_by_instance"` // only available for the iis/webserver metricset at the moment
 }
 
+const metricsetName = "perfmon"
+
 func init() {
-	mb.Registry.MustAddMetricSet("windows", "perfmon", New)
+	mb.Registry.MustAddMetricSet("windows", metricsetName, New)
```
Comment (jsoriano) [inline metricbeat/module/windows/perfmon/perfmon.go]: I am thinking now that as these metricsets don't need a host because they only query perfmon, we could add this:
```suggestion
	mb.Registry.MustAddMetricSet("windows", metricsetName, New,
	                    mb.WithHostParser(parse.EmptyHostParser),
	)
```

File: metricbeat/module/windows/perfmon/perfmon.go (hunk 1 above)
Comment (jsoriano) [reply metricbeat/module/windows/perfmon/perfmon.go]: Oh ok, I would prefer if we can make it more explicit, like with an additional internal flag in the metricset.

Could it happen that we create another metricset based on perfmon where the instance label is also mandatory?

File: metricbeat/module/windows/perfmon/reader.go [hunk 12]
```diff
@@ -137,33 +137,45 @@ func (r *Reader) RefreshCounterPaths() error {
...
 
 	// Get the values.
-	values, err := r.query.GetFormattedCounterValues()
+	values, err := re.query.GetFormattedCounterValues()
 	if err != nil {
 		return nil, errors.Wrap(err, "failed formatting counter values")
 	}
+	var events []mb.Event
+	// GroupAllCountersTo config option will only apply to the webserver metricset, where counters for all instances are aggregated
+	if metricsetName != metricset && re.config.GroupAllCountersTo != "" {
```
Comment (jsoriano) [inline metricbeat/module/windows/perfmon/reader.go]: Similarly here, I would prefer to don't depend on the metricset name for certain features, I would prefer this to be more explicit.
Also this generic metricset shouldn't have something specific for the webserver metricset (or any other metricset) if possible.

File: metricbeat/module/windows/perfmon/perfmon.go [hunk 13]
```diff
@@ -40,13 +40,18 @@ type CounterConfig struct {
...
 type Config struct {
-	IgnoreNECounters  bool            `config:"perfmon.ignore_non_existent_counters"`
-	GroupMeasurements bool            `config:"perfmon.group_measurements_by_instance"`
-	CounterConfig     []CounterConfig `config:"perfmon.counters" validate:"required"`
+	IgnoreNECounters   bool            `config:"perfmon.ignore_non_existent_counters"`
+	GroupMeasurements  bool            `config:"perfmon.group_measurements_by_instance"`
+	CounterConfig      []CounterConfig `config:"perfmon.counters" validate:"required"`
+	GroupAllCountersTo string          `config:"perfmon.group_all_counter"` // only available for the iis/webserver metricset at the moment
+	// could be used in the future if any solutions are built for https://github.com/elastic/beats/issues/16366
+	//FilterByInstance   []string        `config:"perfmon.filter_by_instance"` // only available for the iis/webserver metricset at the moment
```
Comment (jsoriano) [inline metricbeat/module/windows/perfmon/perfmon.go]: Remove commented out code?

File: metricbeat/module/iis/application_pool/reader_test.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+// TestNewReaderValid should successfully instantiate the reader.
+func TestNewReaderValid(t *testing.T) {
+	reader, err := newReader()
+	assert.Nil(t, err)
+	assert.NotNil(t, reader)
+	assert.NotNil(t, reader.Query)
+	assert.NotNil(t, reader.Query.Handle)
+	assert.NotNil(t, reader.Query.Counters)
+	assert.Zero(t, len(reader.Query.Counters))
+	defer reader.close()
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply metricbeat/module/iis/application_pool/reader_test.go]: This comment is unaddressed (not so important in any case).

File: metricbeat/module/windows/perfmon/perfmon.go (hunk 1 above)
Comment (narph) [reply metricbeat/module/windows/perfmon/perfmon.go]: I have decided to leave the instance label optional in the perfmon metricset so no internal flag is required.

File: metricbeat/module/windows/perfmon/reader.go (hunk 5 above)
Comment (narph) [reply metricbeat/module/windows/perfmon/reader.go]: removed the check for the metricset name

File: metricbeat/helper/windows/pdh/pdh_windows.go [hunk 15]
```diff
@@ -51,7 +51,8 @@ type PdhCounterHandle uintptr
 
 var InvalidCounterHandle = ^PdhCounterHandle(0)
 
-const PERF_DETAIL_WIZARD = 400
+// counter detail level
```
[the remark this answers is not in the document]
Comment (narph) [reply metricbeat/helper/windows/pdh/pdh_windows.go]: made the changes

File: metricbeat/module/iis/application_pool/_meta/data.json [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    },
+    "service" : {
+        "type" : "iis"
+    },
+    "iis" : {
+        "application_pool" : {
+            "net_clr" : {
+                "total_exceptions_thrown" : 0
+            },
+            "worker_process_id" : 11956,
```
[the remark this answers is not in the document]
Comment (narph) [reply metricbeat/module/iis/application_pool/_meta/data.json]: thanks for pointing that out

File: metricbeat/module/iis/application_pool/_meta/data.json [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+        "type" : "iis"
+    },
+    "iis" : {
+        "application_pool" : {
+            "net_clr" : {
+                "total_exceptions_thrown" : 0
+            },
+            "name" : "test.local",
+            "process" : {
+                "pid" : 11956,
```
Comment (jsoriano) [inline metricbeat/module/iis/application_pool/_meta/data.json]: I meant to use the ECS `process.pid`, at the root level. But if it complicates things with the permon metricset we can leave moving this field to ECS for the future.

Comment (narph) [reply metricbeat/module/iis/application_pool/_meta/data.json]: @jsoriano , indeed, my mistake, process.pid is a root filed now, application pool does not go through perfmon so that simplified the case.

File: metricbeat/module/iis/application_pool/reader.go [hunk 18]
```diff
@@ -167,7 +170,11 @@ func (re *Reader) fetch(names []string) ([]mb.Event, error) {
 				if val.Instance == appPool.Name {
 					events[appPool.Name].MetricSetFields.Put(appPool.counters[counterPath], val.Measurement)
 				} else if hasWorkerProcess(val.Instance, workers, appPool.WorkerProcessIds) {
-					events[appPool.Name].MetricSetFields.Put(re.WorkerProcesses[counterPath], val.Measurement)
+					if re.WorkerProcesses[counterPath] == ecsProcessId {
```
Comment (jsoriano) [inline metricbeat/module/iis/application_pool/reader.go]: We could also explicitly move the field once the whole event is generated. This is more efficient though :slightly_smiling_face:

File: metricbeat/module/iis/webserver/manifest.yml [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+      query: '\Web Service(_Total)\Total Bytes Sent'
+    - instance_label: ''
+      measurement_label: network.bytes_sent_per_sec
+      query: '\Web Service(_Total)\Bytes Sent/sec'
+    - instance_label: ''
+      measurement_label: network.bytes_received_per_sec
+      query: '\Web Service(_Total)\Bytes Received/sec'
+    - instance_label: ''
+      measurement_label: network.current_connections
+      query: 'Web Service(_Total)\Current Connections'
```
Comment (sorantis) [inline metricbeat/module/iis/webserver/manifest.yml]: Would be great to also have 
`Web Service(*)Current Anonymous Users`
`Web Service(*)Current NonAnonymous Users`
`Web Service(*)Total Anonymous Users`
`Web Service(*)Anonymous Users/sec`
`Web Service(*)Total NonAnonymous Users`

File: metricbeat/module/iis/webserver/manifest.yml [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+      query: '\Web Service(_Total)\Post Requests/sec'
+    - instance_label: ''
+      measurement_label: network.total_delete_requests
+      query: '\Web Service(_Total)\Total Delete Requests'
+    - instance_label: ''
+      measurement_label: network.delete_requests_per_sec
+      query: '\Web Service(_Total)\Delete Requests/sec'
+    - instance_label: ''
+      measurement_label: network.service_uptime
+      query: '\Web Service(_Total)\Service Uptime'
```
Comment (sorantis) [inline metricbeat/module/iis/webserver/manifest.yml]: We can also add error counters:
`Web Service(*)Total Not Found Errors`
`Web Service(*)Not Found Errors/sec`
`Web Service(*)Total Locked Errors`
`Web Service(*)Locked Errors/sec`

File: metricbeat/module/iis/webserver/manifest.yml [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+    #cache
```
Comment (sorantis) [inline metricbeat/module/iis/webserver/manifest.yml]: Also:
`Web Service(*)Total CGI Requests`
`Web Service(*)CGI Requests/sec`
`Web Service(*)Total ISAPI Extension Requests`
`Web Service(*)ISAPI Extension Requests/sec`

Comment (sorantis) [reply metricbeat/module/iis/webserver/manifest.yml]: https://pandorafms.com/library/iis-monitoring/

File: metricbeat/module/iis/webserver/manifest.yml [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+    - instance_label: ''
+      measurement_label: network.total_bytes_sent
+      query: '\Web Service(_Total)\Total Bytes Sent'
+    - instance_label: ''
+      measurement_label: network.bytes_sent_per_sec
+      query: '\Web Service(_Total)\Bytes Sent/sec'
+    - instance_label: ''
+      measurement_label: network.bytes_received_per_sec
+      query: '\Web Service(_Total)\Bytes Received/sec'
+    - instance_label: ''
```
Comment (sorantis) [inline metricbeat/module/iis/webserver/manifest.yml]: Add file related metrics:
`Web Service(*)Total Files Sent`
`Web Service(*)Files Sent/sec`
`Web Service(*)Total Files Received`
`Web Service(*)Files Received/sec`
`Web Service(*)Total Files Transferred`
`Web Service(*)Files/sec`

File: metricbeat/module/iis/application_pool/reader.go [hunk 23]
```diff
@@ -0,0 +1,276 @@
...
+	"process.virtual_bytes":                "\\Process(w3wp*)\\Virtual Bytes",
+	"process.page_faults_per_sec":          "\\Process(w3wp*)\\Page Faults/sec",
+	"process.io_read_operations_per_sec":   "\\Process(w3wp*)\\IO Read Operations/sec",
+	"process.io_write_operations_per_sec":  "\\Process(w3wp*)\\IO Write Operations/sec",
+	"net_clr.total_exceptions_thrown":      "\\.NET CLR Exceptions(w3wp*)\\# of Exceps Thrown",
+	"net_clr.exceptions_thrown_per_sec":    "\\.NET CLR Exceptions(w3wp*)\\# of Exceps Thrown / sec",
+	"net_clr.filters_per_sec":              "\\.NET CLR Exceptions(w3wp*)\\# of Filters / sec",
+	"net_clr.finallys_per_sec":             "\\.NET CLR Exceptions(w3wp*)\\# of Finallys / sec",
+	"net_clr.throw_to_catch_depth_per_sec": "\\.NET CLR Exceptions(w3wp*)\\Throw To Catch Depth / sec",
+}
```
Comment (sorantis) [inline metricbeat/module/iis/application_pool/reader.go]: We should also add .NET specific metrics:

* `ASP.NET Applications(__Total__)\Cache Total Trims`
Total number of entries forcibly removed from
the cache due to memory pressure.
* `ASP.NET Applications(__Total__)\Cache Total Entries`
Total number of entries within the cache (both
internal and user added)
* `ASP.NET Applications(__Total__)\Cache Total Hit Ratio`
Ratio of hits from all cache calls.
* `ASP.NET Applications(__Total__)\Cache Total Turnover Rate`
Number of additions and removals to the total
cache per second.
* `ASP.NET Applications(__Total__)\Output Cache Entries`
Current number of entries in the output cache.
* `ASP.NET Applications(__Total__)\Output Cache Hits`
Total number of output cacheable requests served from the output cache.
* `ASP.NET Applications(__Total__)\Output Cache Hit Ratio`
Ratio of hits to requests for output cacheable requests.
* `ASP.NET Applications(__Total__)\Output Cache Turnover Rate`
Number of additions and removals to the output cache per second.
* `ASP.NET Applications(__Total__)\Compilations Total`
Number of .asax, .ascx, .ashx, .asmx, or .aspx source files dynamically compiled.
* `ASP.NET Applications(__Total__)\Errors Total/Sec`
Rate of errors occurred.
* `ASP.NET Applications(__Total__)\Pipeline Instance Count`
Number of active pipeline instances.
* `ASP.NET Applications(__Total__)\Requests Executing`
The number of requests currently executing.
* `ASP.NET Applications(__Total__)\Requests in Application Queue`
The number of requests in the application request queue.
* `ASP.NET Applications(__Total__)\Requests/Sec` 
The number of requests executed per second.
* `ASP.NET\Application Restarts` 
Number of times the application has been restarted during the Web server's lifetime.
* `ASP.NET\Request Wait Time` 
The number of milliseconds the most recent request was waiting in the queue.

File: metricbeat/module/iis/application_pool/reader.go [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+	"process.cpu_usage_perc":               "\\Process(w3wp*)\\% Processor Time",
+	"process.handle_count":                 "\\Process(w3wp*)\\Handle Count",
+	"process.thread_count":                 "\\Process(w3wp*)\\Thread Count",
+	"process.working_set":                  "\\Process(w3wp*)\\Working Set",
+	"process.private_byte":                 "\\Process(w3wp*)\\Private Bytes",
+	"process.virtual_bytes":                "\\Process(w3wp*)\\Virtual Bytes",
+	"process.page_faults_per_sec":          "\\Process(w3wp*)\\Page Faults/sec",
+	"process.io_read_operations_per_sec":   "\\Process(w3wp*)\\IO Read Operations/sec",
+	"process.io_write_operations_per_sec":  "\\Process(w3wp*)\\IO Write Operations/sec",
+	"net_clr.total_exceptions_thrown":      "\\.NET CLR Exceptions(w3wp*)\\# of Exceps Thrown",
```
Comment (sorantis) [inline metricbeat/module/iis/application_pool/reader.go]: Additional .NET CLR metrics:
* `.NET CLR Memory(w3wp)\% Time in GC`
The percentage of elapsed time that was spent in performing a garbage collection (GC) since the last GC cycle.
* `.NET CLR Memory(w3wp)\# Bytes in all Heaps` 
The sum of four other counters—Gen 0 Heap Size, Gen 1 Heap Size, Gen 2 Heap Size, and the Large Object Heap Size. 
* `.NET CLR Memory(w3wp)\# Total Committed Bytes`
The amount of virtual memory (in bytes) currently committed by the Garbage Collector. 
* `.NET CLR Memory(w3wp)\Process ID` 
The process ID of the CLR process instance being monitored.
* `.NET CLR Security(w3wp)\% Time in RT checks` 
The percentage of elapsed time spent in performing run-time Code Access Security (CAS) checks since the last such check. It represents the last observed value; it’s not an average.

File: metricbeat/module/iis/website/manifest.yml [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+      measurement_label: current_connections
+      query: 'Web Service(*)\Current Connections'
+    - instance_label: 'name'
+      measurement_label: maximum_connections
+      query: '\Web Service(*)\Maximum Connections'
+    - instance_label: 'name'
+      measurement_label: total_connection_attempts
+      query: '\Web Service(*)\Total Connection Attempts (all instances)'
+    - instance_label: 'name'
+      measurement_label: total_get_requests
```
[the remark this answers is not in the document]
Comment (sorantis) [reply metricbeat/module/iis/website/manifest.yml]: `\Web Service(*)\Total PUT Requests`
`\Web Service(*)\PUT Requests/sec`

`\Web Service(*)\Total Head Requests`
`\Web Service(*)\Head Requests/sec`

`\Web Service(*)\Options Requests/sec`
`\Web Service(*)\Total Options Requests`

`\Web Service(*)\Other Request Methods/sec`
`\Web Service(*)\Total Other Request Methods`

`\Web Service(*)\Total Trace Requests`
`\Web Service(*)\Trace Requests/sec`

`\Web Service(*)\Total Unlock Requests`
`\Web Service(*)\Unlock Requests/sec`

File: metricbeat/module/iis/application_pool/reader.go (hunk 24 above)
Comment (narph) [reply metricbeat/module/iis/application_pool/reader.go]: @sorantis , adding almost all the counters suggested in all 3 metricsets more than doubled the CPU usage, do you mind going over the lists again and picking just a few that could be of interest.

Comment (sorantis) [reply metricbeat/module/iis/application_pool/reader.go]: skip

File: metricbeat/module/iis/webserver/manifest.yml (hunk 22 above)
Comment (sorantis) [reply metricbeat/module/iis/webserver/manifest.yml]: skip

File: metricbeat/module/iis/application_pool/reader.go (hunk 23 above)
Comment (sorantis) [reply metricbeat/module/iis/application_pool/reader.go]: skip all except
```
ASP.NET Applications(__Total__)\Errors Total/Sec
ASP.NET Applications(__Total__)\Pipeline Instance Count
ASP.NET Applications(__Total__)\Requests Executing
ASP.NET Applications(__Total__)\Requests in Application Queue
ASP.NET Applications(__Total__)\Requests/Sec
ASP.NET\Application Restarts
ASP.NET\Request Wait Time
```
